inc/sitemaps/class-sitemaps.php

```diff
@@ -207,15 +194,51 @@ public function redirect( $query ) {
 			return;
 		}
 
-		if ( $caching && ! $this->transient ) {
-			$this->cache->store_sitemap( $type, $this->current_page, $this->sitemap );
-		}
-
 		$this->output();
 		$this->sitemap_close();
 	}
 
 	/**
+	 * Try to get the sitemap from cache
+	 *
+	 * @param string $type        Sitemap type.
+	 * @param int    $page_number The page number to retrieve.
+	 *
+	 * @return bool If the sitemap has been retrieved from cache.
+	 */
+	private function get_sitemap_from_cache( $type, $page_number ) {
+		if ( true !== $this->cache->is_enabled() ) {
+			return false;
+		}
+
+		/**
+		 * Fires before the attempt to retrieve XML sitemap from the transient cache.
+		 *
+		 * @param WPSEO_Sitemaps $sitemaps Sitemaps object.
+		 */
+		do_action( 'wpseo_sitemap_stylesheet_cache_' . $type, $this );
+
+		$sitemap_cache_data = $this->cache->get_sitemap_data( $type, $page_number );
+
+		// No cache was found, build it and save to cache.
+		if ( empty( $sitemap_cache_data ) ) {
+			$this->build_sitemap( $type );
+
+			$this->cache->store_sitemap( $type, $page_number, $this->sitemap, ! $this->bad_sitemap );
```

Comment:
For single responsibility, it might be best to extract the caching to another function, as this function reads `get_sitemap_from_cache`